FWCore/ParameterSet/python/Config.py

```diff
@@ -1431,6 +1442,45 @@ def validate(self):
         #    raise RuntimeError("No input source was found for this process")
         pass
 
+    def handleProcessAccelerators(self, parameterSet):
+        allAccelerators = set([label for acc in self.__dict__['_Process__accelerators'].values() for label in acc.labels()])
+        # 'cpu' accelerator is always implicitly there
+        availableAccelerators = set(["cpu"])
+        for acc in self.__dict__['_Process__accelerators'].values():
+            for l in acc.enabledLabels():
+                availableAccelerators.add(l)
+        unavailableAccelerators = allAccelerators.difference(availableAccelerators)
+        availableAccelerators = sorted(list(availableAccelerators))
+        parameterSet.addVString(False, "@available_accelerators", availableAccelerators)
+
+        # Resolve wildcards
+        if "*" in self.options.accelerators:
+            if len(self.options.accelerators) >= 2:
+                raise ValueError("process.options.accelerators may contain '*' only as the only element, now it has {} elements".format(len(self.options.accelerators)))
+            self.options.accelerators = availableAccelerators
```

Comment:
just an idea, do you think it would be useful to store the original list of requested accelerator (_e.g._ in `"@requested_accelerators"`) to give a more meaningful error message in `ensureAvailableAccelerators()` ?

Reply:
Possibly, although then I'd change the mechanism to not rewrite `process.options.accelerators` but pass the set of selected accelerators `ProcessAccelerator.apply()` in an explicit argument (which could be more clear anyway).

Reply:
The original list of patterns is now passed to C++ code and included in the exception message.